8307160: Fix AWT/2D/A11Y to support the permissive- flag on the Microsoft Visual C compiler

TheShermanTanker opened the pull request:
We should set the -permissive- flag for the Microsoft Visual C compiler, as was requested by the now backed out [JDK-8241499](https://bugs.openjdk.org/browse/JDK-8241499). Doing so makes the Visual C compiler much less accepting of ill formed code, which will improve code quality on Windows in the future.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (3 reviews required, with at least 3 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8307160](https://bugs.openjdk.org/browse/JDK-8307160): Fix AWT/2D/A11Y to support the permissive- flag on the Microsoft Visual C compiler (**Enhancement** - P4) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [51230f3d](https://git.openjdk.org/jdk/pull/15096/files/51230f3d0eb942b18f5d951891406aba3c91dac7)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [51230f3d](https://git.openjdk.org/jdk/pull/15096/files/51230f3d0eb942b18f5d951891406aba3c91dac7)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15096/head:pull/15096` \
`$ git checkout pull/15096`

Update a local copy of the PR: \
`$ git checkout pull/15096` \
`$ git pull https://git.openjdk.org/jdk.git pull/15096/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15096`

View PR using the GUI difftool: \
`$ git pr show -t 15096`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15096.diff">https://git.openjdk.org/jdk/pull/15096.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15096#issuecomment-1659450755)

TheShermanTanker commented:
That broken mismatched exception specifier in allocation.cpp has been annoying the absolute \*\*\*\* out of me for a long time now, feels good to see it go

TheShermanTanker commented:
/label hotspot

dholmes-ora commented on src/hotspot/os/windows/symbolengine.cpp:
```diff
@@ -111,7 +111,7 @@ class SimpleBufferWithFallback {
       _capacity = (int)(sizeof(_fallback_buffer) / sizeof(T));
     }
-    _p[0] = '\0';
+    _p[0] = (T) '\0';
```
I find this use of `'\0'` quite suspicious in itself, but it is consistent with the use of `'X'` in `imprint_sentinel()`. I'm unclear what types are possible for `T` in this code.

TheShermanTanker replied:
There are currently only 2 possible types of HMODULE and char/uint8_t for T at the moment. Weirdly enough only line 126 errors out without the cast while line 114, despite having the same problem, doesn't, but I added the cast to both lines for consistency. If someone else knows why I could probably deal with this code in a better way besides just casting it to T (which I also am reluctant to do)

TheShermanTanker replied:
I just checked and the value of the sentinel is ultimately the prvalue 88. I don't know if we'd want to replace all the weird char usages here with explicit values of 0 (and 88 for the sentinel). Maybe future reviews can help with that

tstuefe replied:
I wrote this code ages ago. I'm not sure what's weird or suspicious about it, though. The comment at the file's beginning explains this code's motivation.

The buffer was never thought to be used for something different than HANDLEs or characters, where the assignment of integer literals work. I often use char constants for sentinels as debugging aid. As for `'\0'`, that indicates to the casual code reader that this is a termination of a string, better than had I used a plain 0.

dholmes-ora replied:
Because there is nothing to state what T may be, I found assigning character literals to be odd. If T is char and the buffer is meant to be a C string then it makes more sense. But for non-char T it just raised questions for me.

tstuefe replied:
I see that. We can remove the sentinel stuff, which leaves us with the zero termination. Arguably, this could be done by sub classes that derive from the char instantiation.

dholmes-ora commented on src/hotspot/share/memory/allocation.cpp:
```diff
@@ -111,7 +111,7 @@ void* ArenaObj::operator new(size_t size, Arena *arena) throw() {
 //
 
-void* AnyObj::operator new(size_t size, Arena *arena) throw() {
+void* AnyObj::operator new(size_t size, Arena *arena) {
```
Please remind us what the issue is with `throw()` as this is a change to shared code and it looks very inconsistent to remove it only for this one definition of `operator new`.

TheShermanTanker replied:
Hi David, this throw was removed for this method's declaration in the corresponding hpp file in an earlier commit, but the Committer forgot to remove the throw() from the definition as well (the error can be viewed in the GHA for the earlier versions of this PR). I believe gcc only errors on this in C++17 mode (for some weird reason) and not our current C++14 mode, which is why it gets a pass on other platforms

dholmes-ora replied:
Thanks for that.

TheShermanTanker commented:
/label client security

TheShermanTanker commented:
Ready for initial rounds of review

dholmes-ora reviewed (approved):
Hotspot and build changes seem okay.

Thanks.

dholmes-ora commented:
/reviewers needed 2

TheShermanTanker commented:
Setting reviewers required to 4 for each subcomponent

TheShermanTanker commented:
/reviewers 4

erikj79 reviewed (approved):
Build change looks ok.

tstuefe reviewed (commented):
Hi Julian,

Cursory review. I stopped halfway in when I noticed that I had no idea why many of the changes were done. They did not seem to have an obvious connection to the title of the JBS.

Cheers, Thomas

tstuefe commented:
I think I see now that the added scopes were to prevent variable life scopes from intersecting goto's?

TheShermanTanker commented on src/java.desktop/windows/native/libawt/windows/awt_DnDDS.cpp:
```diff
@@ -39,10 +39,13 @@ void * operator new(size_t size) {return operator new(size, "stl", 1);}
+#include "awt_DCHolder.h"
+
 #include "jlong.h"
+#include "awt.h"
```
This is a weird one, but in awt we #define malloc Do_Not_Use_Malloc... And so on. Without this reordering awt_ole.h (which includes comdef.h) also uses the redefined malloc somewhere (I could not find where in comip.h that it's used) which breaks compilation. I do find it strange that without permissive- it doesn't break though. Same applies to the other reordering of includes

TheShermanTanker replied:
To my knowledge this has been an issue before, and was fixed by [8241087](https://bugs.openjdk.org/browse/JDK-8241087), but it missed this one. The code that needs the malloc is in a template in Microsoft's headers, which I don't think we can change. I'm not sure what other solution we can come up with for this

TheShermanTanker commented:
/reviewers 4, reviewer

TheShermanTanker commented:
> I think I see now that the added scopes were to prevent variable life scopes from intersecting goto's?

Yes, that's why

tstuefe commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -2884,22 +2884,22 @@ LONG WINAPI topLevelVectoredExceptionFilter(struct _EXCEPTION_POINTERS* exceptio
-  if (InterceptOSException) goto exit;
-  DWORD exception_code = exceptionInfo->ExceptionRecord->ExceptionCode;
+  if (!InterceptOSException) {
+    DWORD exception_code = exceptionInfo->ExceptionRecord->ExceptionCode;
```
Ah, okay. Thanks for clarifying.

tstuefe commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1577,63 +1580,65 @@ void AwtFrame::_NotifyModalBlocked(void *param)
             }
         }
     }
+
```
A possible improvement later (and for a future RFE) would be to use RAII for deletion and then get rid of the labels. awt is one of the few places that uses C++ for native code, so why not.

magicus replied:
The comment seems answered. Please resolve this discussion.

tstuefe commented on src/java.desktop/windows/native/libawt/windows/awt_TextComponent.cpp:
```diff
@@ -53,12 +53,14 @@ struct EnableEditingStruct {
+jmethodID AwtTextComponent::canAccessClipboardMID;
+AwtTextComponent::OleCallback AwtTextComponent::sm_oleCallback;
+WNDPROC AwtTextComponent::sm_pDefWindowProc = NULL;
+
```
Did the compiler complain here? I'm fine with the change, just wanted to know the reason.

TheShermanTanker replied:
the latter two are inside an extern "C" block, meaning their initial C++ linkage (by virtue of them being static class members) conflicts with the now C linkage, also the comment there states the AwtComponent fields are supposed to be set here, and I have no idea why this was not done, so I moved them all to be under that comment

magicus replied:
The question seem answered. Please resolve the discussion.

TheShermanTanker commented:
Bumping, anything from security and awt folk?

TheShermanTanker commented:
/reviewers 5 reviewer

prrace commented:
I have no time to look at the client changes for quite some time so do not push it. No matter how many other people approve it.
And in the meantime you can 
(1) explain how many client tests you ran - and it had better be all of them :-)
(2) add a comment in the PR to explain each client change you made - at the site of the change.
  Comments like "Fields in awt_TextComponent.cpp" are not an explanation
(3) On the broader front, I note that you just start off with "We should set the -permissive- flag for the Microsoft Visual C compiler"
without explaining WHY and what it does. In fact from what I've read about -it seems like we would not want it .. it downgrades errors to warnings .. and surely that's a bad thing ?

dean-long commented:
I had the same confusion, reading "-permissive-" the same as /permissive, but the trailing "-" means disable.

TheShermanTanker commented:
> I have no time to look at the client changes for quite some time so do not push it. No matter how many other people approve it. And in the meantime you can (1) explain how many client tests you ran - and it had better be all of them :-) (2) add a comment in the PR to explain each client change you made - at the site of the change. Comments like "Fields in awt_TextComponent.cpp" are not an explanation (3) On the broader front, I note that you just start off with "We should set the -permissive- flag for the Microsoft Visual C compiler" without explaining WHY and what it does. In fact from what I've read about -it seems like we would not want it .. it downgrades errors to warnings .. and surely that's a bad thing ?

No worries, I was not planning on committing without approval from each group, this was why I set the required reviewers to 5, 2 for HotSpot, one for build, and so on. I'll add the comments later on, but right now I can explain the changes directly since it's slightly easier to do so:

The changes to awt involve enclosing several areas which make heavy use of goto in scopes to avoid them jumping over the initialization of locals, reordering the order of includes to avoid Microsoft specific headers from using our redefined malloc, moving the definition of C++ static fields outside of extern "C" blocks, and removing the class qualifiers from field declarations in several classes. The last one is fairly obvious - It isn't legal C++ grammar to do so. For the fields that were moved, they have C++ linkage initially by virtue of being static members of a class, but their definitions are inside extern "C" blocks, which causes their linkage to conflict between C++ and C. For the reordering of includes, see my conversation with Thomas - With permissive- enabled C++ templates are properly conforming, and the header comip.h (which is included by comdef.h) uses a templated malloc somewhere, which due to our #define malloc in alloc.h fails since the identifier is obviously not defined. This is actually a major problem that is potentially causing an invisible bug in awt.dll somewhere, aside from it not compiling with permissive-, the redefining of malloc and all other allocators like this is something @tstuefe has expressed some concern over, but I don't really know how to deal with the bigger underlying problem outside of reordering the includes to not let comdef.h use our #define'd malloc. For the enclosing of gotos in scopes Thomas has suggested using RAII instead of labels in awt, which is something that could be done outside of this change, but I'm still figuring out if doing so for the sites that this is a problem for in this change would be neater

Noted, I'll also improve the explaination on what permissive- does and why we want it in just a moment

mrserb commented:
Why initial patch was reverter using this comment: "unfortunately, did not prove to be as useful as expected"? What was the problem? What about the "pedantic" option which was added last time?

TheShermanTanker commented:
> Why initial patch was reverter using this comment: "unfortunately, did not prove to be as useful as expected"? What was the problem? What about the "pedantic" option which was added last time?

Hi, the problem was that gcc and clang are by default extremely strict with regards to conforming to C++, and the -pedantic flag simply made them so strict that it was impossible to compile the JDK with them. In an ideal world we'd be able to compile with -pedantic enabled, but that seems to be a lot of work required before that can be achieved, so it was reverted. In contrast, Visual C compiler is much too lax with what it allows you to get away with by default, so the permissive- flag is needed to make it match clang and gcc's default strictness. Additionally, permissive- is going to become the default on Visual C in the future, so we should add this as a flag to allow the JDK to be prepared when that time arrives

TheShermanTanker commented:
Adding extra reviewer for jdk.accessibility
/reviewers 6 reviewer

TheShermanTanker commented:
Bumping, anyone from security?

TheShermanTanker commented:
Bumping

TheShermanTanker commented:
Bumping?

TheShermanTanker commented:
Pinging

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_DnDDT.cpp:
```diff
@@ -27,11 +27,11 @@
 #include "java_awt_dnd_DnDConstants.h"
 #include "sun_awt_windows_WDropTargetContextPeer.h"
 #include "awt_Container.h"
-#include "awt_ole.h"
```
Is this related to the `#define malloc Do_Not_Use_Malloc` issue? If so, the required ordering of includes should be documented here as well.

TheShermanTanker replied:
Will do

magicus commented:
@TheShermanTanker In my experience, getting reviews from all areas for issues like this that cuts through the entire JDK can be difficult. Another approach, which requires more work from your side, but hopefully less from the reviewers' (and thus makes it easier for them to review) is to split this PR into multiple ones: One for each area (basically, area == mailing list) that just makes the changes to the code necessary to (in the future) turn on /permissive-. And then finally a small "finishing" PR which just touches the makefile and enables the flag, when all code is fixed. 

As a side effect, it is also 100% clear that all parts of the code has been correctly reviewed, since then reviewers do not need to leave conditions on their reviews ("i only looked at the foo parts").

TheShermanTanker commented:
> @TheShermanTanker In my experience, getting reviews from all areas for issues like this that cuts through the entire JDK can be difficult. Another approach, which requires more work from your side, but hopefully less from the reviewers' (and thus makes it easier for them to review) is to split this PR into multiple ones: One for each area (basically, area == mailing list) that just makes the changes to the code necessary to (in the future) turn on /permissive-. And then finally a small "finishing" PR which just touches the makefile and enables the flag, when all code is fixed.
> 
> As a side effect, it is also 100% clear that all parts of the code has been correctly reviewed, since then reviewers do not need to leave conditions on their reviews ("i only looked at the foo parts").

I understand, will split this into multiple changes after I answer all queries above

TheShermanTanker commented:
closing and deleting for now

TheShermanTanker commented:
Reopening - I will use this as the portion of the change for java.desktop and jdk.accessibility

TheShermanTanker commented:
/label remove hotspot security

TheShermanTanker commented:
/reviewers 1 reviewer

magicus commented:
This got a bit messy. In hindsight it would probably have been better to leave this closed and open a new bug. But what's done is done, let's not mess around any more at this point.

If all other places where `permissive-` failed are fixed, then I think it is okay to include the flag in this PR. Otherwise, you cannot do that. As a general rule, changing compiler options is best done in a separate patch, to differentiate any potential bad effect from actual code changes. However, in this case, I don't think the flag will cause any differences in the generated code, only in what syntax the compiler accepts, so I guess it is fine.

/reviewers 2

TheShermanTanker commented:
> This got a bit messy. In hindsight it would probably have been better to leave this closed and open a new bug. But what's done is done, let's not mess around any more at this point.
> 
> If all other places where `permissive-` failed are fixed, then I think it is okay to include the flag in this PR. Otherwise, you cannot do that. As a general rule, changing compiler options is best done in a separate patch, to differentiate any potential bad effect from actual code changes. However, in this case, I don't think the flag will cause any differences in the generated code, only in what syntax the compiler accepts, so I guess it is fine.
> 
> /reviewers 2

Sorry I was busy with work on my end, I'm not planning to integrate -permissive- in this Pull Request, I must have missed it. I'll remove it soon

magicus commented:
/label -build

TheShermanTanker commented:
I am only adding permissive- to check for any flaws I have left in there, rest assured I will not be integrating the permissive- flag in this changeset

tstuefe commented:
I'm removing myself from this. There are enough eyes on this.

prrace commented:
/reviewers 3 reviewers

dean-long commented on src/java.desktop/windows/native/libawt/windows/awt_Canvas.cpp:
```diff
@@ -214,7 +214,8 @@ void AwtCanvas::_SetEraseBackground(void *param)
     JNI_CHECK_PEER_GOTO(canvas, ret);
 
-    AwtCanvas *c = (AwtCanvas*)pData;
+    AwtCanvas *c;
+    c = (AwtCanvas*) pData;
     c->m_eraseBackground = doErase;
     c->m_eraseBackgroundOnResize = doEraseOnResize;
```
```suggestion
    {
        AwtCanvas *c = (AwtCanvas*) pData;
        c->m_eraseBackground = doErase;
        c->m_eraseBackgroundOnResize = doEraseOnResize;
    }
```
Does wrapping in {} work?  I think it looks better, and that's how we handle case labels in switch statements, isn't it?

TheShermanTanker replied:
Wrapping it in a scope does work, but the problem is that it unfortunately becomes very messy and hard to review when the scope spans over a lot of lines (see below for instance, where I haven't switched from a scope to the split declaration and assignment yet). This approach was suggested by @djelinski in the review of https://github.com/openjdk/jdk/pull/15996

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -57,6 +57,8 @@
 #define ROUND_TO_INT(num)     ((int) floor((num) + 0.5))
 
+jfieldID AwtPrintDialog::pageID;
+
```
where and why did this come from ?

TheShermanTanker replied:
This came from below, all I did was move it up and out of the extern "C" block. This cannot be inside extern "C" because it is a static class member and has permanent C++ name mangling

magicus replied:
The question seems answered. Please resolve this discussion.

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -528,184 +529,186 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
-        if ((setup.hDevMode == NULL) && (setup.hDevNames == NULL)) {
-            doIt = JNI_FALSE;
-            goto done;
+        HWND parentID = AwtPrintControl::getParentID(env, self);
```
diff is making this really hard to compare line by line
So here you've just put the code in { .. } and indented ? 
What is the warning that motivates this ?
I don't see where you describe it.

TheShermanTanker replied:
I haven't gotten around to converting this to the new approach of splitting declaration and assignment, but yes, that is all I've done. I've described the problem here above

TheShermanTanker commented on src/java.desktop/windows/native/libawt/windows/awt_Canvas.cpp:
```diff
@@ -214,7 +214,8 @@ void AwtCanvas::_SetEraseBackground(void *param)
 
-    AwtCanvas *c = (AwtCanvas*)pData;
+    AwtCanvas *c;
+    c = (AwtCanvas*) pData;
```
The problem is that the goto jumps over the initialization of these locals, which C++ does not allow. However, a split declaration and assignment can be jumped over by goto if the local is a primitive type, hence this change is needed

prrace replied:
Really I hate this change. It is wrong.  Well I could go further than that. I have no idea why C++ thinks one is OK and the other is not. Please show the EXACT WARNING FROM THE COMPILER. (Actually why aren't you showing us what the problems are ? )  I don't know if I want to change C++ or something else. Actually C++ is a problem in itself, but that's a personal opinion. If goto is the specific problem and the only thing we can change then we should eliminate goto.

prrace commented:
For everyone of these changes, I want to see what the compiler says if you don't make the change. Else I am flying half blind.

TheShermanTanker commented:
Right, I can simply remove the changes and let the permissive- in the GitHub Actions do the talking

TheShermanTanker commented:
Actually, I can do you one better, now that I think of it, so I don't have to explains potentially deeply nested and confusing error logs:

```
int main() {
    goto skip;
    int i = 0;
    skip:
    std::printf("Done\n");
}
```

C:\Users\vertig0\Downloads>cl.exe -nologo -std:c++14 -permissive- -Fo:goto.o goto.cpp

goto.cpp(5): error C2362: initialization of 'c' is skipped by 'goto skip'
goto.cpp(6): note: see declaration of 'skip'

You simply cannot jump over initialization of a local with a goto in C++, as shown by the Visual C compiler above, does this answer your question?

TheShermanTanker commented:
https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2362?view=msvc-170

prrace commented:
> Actually, I can do you one better, now that I think of it, so I don't have to explains potentially deeply nested and confusing error logs:
> 
> ```
> int main() {
>     goto skip;
>     int i = 0;
>     skip:
>     std::printf("Done\n");
> }
> ```
> 
> C:\Users\vertig0\Downloads>cl.exe -nologo -std:c++14 -permissive- -Fo:goto.o goto.cpp
> 
> goto.cpp(5): error C2362: initialization of 'c' is skipped by 'goto skip' goto.cpp(6): note: see declaration of 'skip'
> 
> You simply cannot jump over initialization of a local with a goto in C++, as shown by the Visual C compiler above, does this answer your question?

It tells me 2 things
(1) C++ is awful, because the change you made to silence it is a no-op
(2) the changes you are making are really wrong, because C++ is telling us it does not like goto, and you aren't addresssing that

TheShermanTanker commented:
> > Actually, I can do you one better, now that I think of it, so I don't have to explains potentially deeply nested and confusing error logs:
> > ```
> > int main() {
> >     goto skip;
> >     int i = 0;
> >     skip:
> >     std::printf("Done\n");
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > C:\Users\vertig0\Downloads>cl.exe -nologo -std:c++14 -permissive- -Fo:goto.o goto.cpp
> > goto.cpp(5): error C2362: initialization of 'c' is skipped by 'goto skip' goto.cpp(6): note: see declaration of 'skip'
> > You simply cannot jump over initialization of a local with a goto in C++, as shown by the Visual C compiler above, does this answer your question?
> 
> It tells me 2 things (1) C++ is awful, because the change you made to silence it is a no-op (2) the changes you are making are really wrong, because C++ is telling us it does not like goto, and you aren't addresssing that

I made the decision to use no-op changes based on the review in https://github.com/openjdk/jdk/pull/15996 at @djelinski's suggestion, because I assumed the best way forward is to make the changes minimally disruptive to client code, is that not what we're aiming for here? The quote-on-quote correct ways to fix this are either to wrap everything in a scope, or rewrite a significant portion of the AWT codebase to remove the gotos, the former is very messy and hard to review (as seen above in one of your earlier reviews), the latter would be difficult to do. Not sure which one you'd prefer?

prrace commented:
> > > Actually, I can do you one better, now that I think of it, so I don't have to explains potentially deeply nested and confusing error logs:
> > > ```
> > > int main() {
> > >     goto skip;
> > >     int i = 0;
> > >     skip:
> > >     std::printf("Done\n");
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > C:\Users\vertig0\Downloads>cl.exe -nologo -std:c++14 -permissive- -Fo:goto.o goto.cpp
> > > goto.cpp(5): error C2362: initialization of 'c' is skipped by 'goto skip' goto.cpp(6): note: see declaration of 'skip'
> > > You simply cannot jump over initialization of a local with a goto in C++, as shown by the Visual C compiler above, does this answer your question?
> > 
> > 
> > It tells me 2 things (1) C++ is awful, because the change you made to silence it is a no-op (2) the changes you are making are really wrong, because C++ is telling us it does not like goto, and you aren't addresssing that
> 
> I made the decision to use no-op changes based on the review in #15996 at @djelinski's suggestion, because I assumed the best way forward is to make the changes minimally disruptive to client code, is that not what we're aiming for here? The quote-on-quote correct ways to fix this are either to wrap everything in a scope, or rewrite a significant portion of the AWT codebase to remove the gotos, the former is very messy and hard to review (as seen above in one of your earlier reviews), the latter would be difficult to do. Not sure which one you'd prefer?
> 
> Side note: What about the jdk.accessibility changes? Are those ok at least?

So this is "illegal" 

 int main() {
  goto skip;
int i = 0;
 skip:
std::printf("Done\n");
 }

and this is legal ?

 int main() {
  goto skip;
 int i;
 i= 0;
 skip:
std::printf("Done\n");
 }

Honestly, I don't like these changes at all.
Anyone looking at the code after the proposed change would just be itching to undo it unless they knew exactly why it was like that.
I'd prefer to leave the code alone and complain to Microsoft as I can't believe they've interpreted / implemented the C++ spec correctly. 
2nd choice is to re-write without gotos.

prrace commented:
PS, yes the A11Y changes are OK.

TheShermanTanker commented:
@prrace I've implemented an RAII strategy in awt_Canvas.cpp as an initial experiment to replace the gotos, if you like it I'll convert the rest to RAII as well

TheShermanTanker commented:
By the way, how outdated is the comment that awt.dll should not rely on msvcp.dll? As far as I can tell we now always distribute msvcp.dll in the JDK, and it would help a lot if I could use the C++ Standard Library for this

TheShermanTanker commented:
@prrace By the way, how outdated is the comment that awt.dll should not rely on msvcp.dll? As far as I can tell we now always distribute msvcp.dll in the JDK, and it would help a lot if I could use the C++ Standard Library for this. Or could I statically link the CRT into awt.dll instead, if awt.dll really should not rely on msvcp.dll?

prrace commented:
> By the way, how outdated is the comment that awt.dll should not rely on msvcp.dll? As far as I can tell we now always distribute msvcp.dll in the JDK, and it would help a lot if I could use the C++ Standard Library for this

Do we ? I'm not sure why that is. We've previously tried hard to avoid that.
Eg https://hg.openjdk.org/jdk9/jdk9/rev/f3c96aea372d

We should still stay away from it.

But it doesn't matter because the changes you are making aren't going in the right direction.
The root of the compiler complaints is the gotos isn't it ?
So the code needs to be changed to not use gotos.
But there are hundreds of those. I guess only a few trigger the warnings ?
So your awt_Canvas example should end up looking something like this
WARNING: I just edited in place, haven't built it, haven't tested it, it is to show what the code should look like
`
void AwtCanvas::_SetEraseBackground(void *param)
{
    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
                    
    SetEraseBackgroundStruct *sebs = (SetEraseBackgroundStruct *)param;
    jobject canvas = sebs->canvas;
                    
    if (canvas == NULL) {
        env->ExceptionClear(); 
        JNU_ThrowNullPointerException(env, "canvas");
    } else {            
        PDATA pData = JNI_GET_PDATA(canvas);
        if (pData == NULL) {
            env->DeleteGlobalRef(canvas);
            THROW_NULL_PDATA_IF_NOT_DESTROYED(canvas);
        } else {            
            AwtCanvas *c = (AwtCanvas*)pData;
            c->m_eraseBackground = sebs->doErase;
            c->m_eraseBackgroundOnResize = sebs->doEraseOnResize;
            env->DeleteGlobalRef(canvas);
        }                   
    }                        
    delete sebs;            
}
`

TheShermanTanker commented:
> But it doesn't matter because the changes you are making aren't going in the right direction.
> The root of the compiler complaints is the gotos isn't it ?
> So the code needs to be changed to not use gotos. I guess you did do that but not in a way I'd endorse

Indeed, I am planning to rewrite it in a way that doesn't use gotos (at least for the areas here where the compiler cannot compile). The reason I asked if the C++ Standard Library could be used is because std::unique_ptr can be used here to very cleanly replace the gotos, with minimal and non-disruptive code changes that would be easy to review, compared to the clunky changes I had to make here (without the C++ Library)

> Do we ? I'm not sure why that is. We've previously tried hard to avoid that. Eg https://hg.openjdk.org/jdk9/jdk9/rev/f3c96aea372d
> 
> We should still stay away from it.

So the dynamic version of the C++ Standard Library (msvcp140.dll) cannot be used. Alright. What if we statically link the C Runtime Libraries to awt.dll instead? This way, not only do we not have to worry all the time about introducing a dependency to msvcp140.dll (as everything is statically linked), but std::unique_ptr also becomes available to use here, and we also wouldn't need to do the std::bad_alloc hack that we currently do too. This is as simple as passing the -MT compiler flag to BUILD_LIBAWT in the Makefile

TheShermanTanker commented:
Bumping

TheShermanTanker commented:
Bumping, @prrace?

prrace commented:
I don't think we should introduce any std:: library support to resolve this.
I'm also fairly sure you can't static link those libraries any more.
The pattern I showed is much cleaner and that's what I want to see here.

TheShermanTanker commented:
> I don't think we should introduce any std:: library support to resolve this. I'm also fairly sure you can't static link those libraries any more. The pattern I showed is much cleaner and that's what I want to see here.

It is still possible to link them statically, we do indeed statically link those libraries in other places too, for instance in jdk.accessibility, see https://github.com/openjdk/jdk/blob/d9a89c59daa40fdc8da620940d5c518a9f18bc7b/make/modules/jdk.accessibility/Lib.gmk#L44

TheShermanTanker commented:
I happened to ask around on the build-dev mailing lists about whether we include msvcp.dll with the JDK, here is Erik's response:

> Back in JDK 8 when we used Visual Studio 2010, we used to not ship
> msvcp*.dll. This changed when I added support for building with Visual
> Studio 2013 [1] in JDK 9. In the patch for that bug I found this text:
> 
> +    # If building with Visual Studio 2010, we can still use
> _STATIC_CPPLIB to
> +    # avoid bundling msvcpNNN.dll. Doesn't work with newer versions of
> visual
> +    # studio.
> 
> So since we switched to Visual Studio 2013, we started to bundle
> msvcp*.dll. It was only ever possible to not bundle it if you built with
> Visual Studio 2010 (or older I suppose). It's pretty safe to say that
> the current mainline JDK build requires msvcp.dll to be bundled. Looking
> at the configure logic, configure will fail unless we find it and the
> copying in open/make/modules/java.base/Copy.gmk is unconditional.
> 
> /Erik
> 
> [1] https://bugs.openjdk.org/browse/JDK-8042707

This may be unrelated, but I really think we should kill the std::bad_alloc hack in awt.dll as such

prrace commented:
> I happened to ask around on the build-dev mailing lists about whether we include msvcp.dll with the JDK, here is Erik's response:

Yes, I saw that. It doesn't make any difference.
Although BTW the comment points out we shouldn't be doing static linking anymore.


> This may be unrelated, but I really think we should kill the std::bad_alloc hack in awt.dll as such

seems unrelated, yes, but I'd have to look at that, to see if there's a reason to touch it. Itchy fingers are  not a valid reason.

TheShermanTanker commented:
> > I happened to ask around on the build-dev mailing lists about whether we include msvcp.dll with the JDK, here is Erik's response:
> 
> Yes, I saw that. It doesn't make any difference. Although BTW the comment points out we shouldn't be doing static linking anymore.
> 
> > This may be unrelated, but I really think we should kill the std::bad_alloc hack in awt.dll as such
> 
> seems unrelated, yes, but I'd have to look at that, to see if there's a reason to touch it. Itchy fingers are not a valid reason.

I've converted awt_Canvas.cpp to use manual cleaup, as requested. Are we sure we really want to do this for the rest of the places where goto is an issue and not rely on the C++ Library? There are multiple usages of JNI\_CHECK\_PEER\_GOTO in those areas, and manually expanding the macro in those areas will quickly become very messy

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Canvas.cpp:
```diff
@@ -202,23 +202,34 @@ MsgRouting AwtCanvas::HandleEvent(MSG *msg, BOOL synthetic)
+    if (canvas == NULL) {
+        env->ExceptionClear();
+        JNU_ThrowNullPointerException(env, "peer");
+        env->DeleteGlobalRef(canvas);
```
you didn't code this up the way I did in my comment, which I think you should have done, and here you are deleting a ref to NULL, and the string passed to throw doesn't mention the canvas, which it can do unlike the old code.

TheShermanTanker replied:
Sorry, I was trying to keep as close to the original code as possible. Woah, does that mean the original was bugged, since it was deleting a NULL canvas? Anyway, should I change the string to canvas or just keep it as "peer" like the original code does in that case?

prrace replied:
Not buggy per se, since NULL is allowed
https://docs.oracle.com/en/java/javase/21/docs/specs/jni/functions.html#deleteglobalref

But that doesn't make it good practice.

Yes, change the name because then it is more precise.

TheShermanTanker replied:
Is the new code better?

TheShermanTanker commented:
@prrace Finally finished converting all of them. Is the new style of change ok?

TheShermanTanker commented:
Re-adding build label since this now touches build flags again
/label build

TheShermanTanker commented on make/autoconf/flags-cflags.m4:
```diff
@@ -562,8 +562,8 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_HELPER],
   elif test "x$TOOLCHAIN_TYPE" = xmicrosoft; then
     # The -utf-8 option sets source and execution character sets to UTF-8 to enable correct
     # compilation of all source files regardless of the active code page on Windows.
-    TOOLCHAIN_CFLAGS_JVM="-nologo -MD -Zc:preprocessor -Zc:strictStrings -Zc:inline -permissive- -utf-8 -MP"
```
@magicus -permissive- automatically turns on -Zc:strictStrings, so specifying it manually becomes redundant when -permissive- is also specified
https://learn.microsoft.com/en-us/cpp/build/reference/zc-strictstrings-disable-string-literal-type-conversion?view=msvc-170

> The -Zc:strictStrings option is off by default. The [-permissive-](https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170) compiler option implicitly sets this option, but it can be overridden by using -Zc:strictStrings-

magicus replied:
I see. Then I guess this is okay. If/when you are finally ready for integration, and turning on -permissive-, let me know so I can take the patch for a spin in our CI systems before integration.

TheShermanTanker commented:
Do not.

TheShermanTanker commented:
Bumping

magicus commented:
What is remaining to get this PR committable? It has such a long history that it is hard to get a grasp on the remaining issues. 

@TheShermanTanker Could you perhaps summarize the remaining hurdles?

TheShermanTanker commented:
> What is remaining to get this PR committable? It has such a long history that it is hard to get a grasp on the remaining issues.
> 
> @TheShermanTanker Could you perhaps summarize the remaining hurdles?

It's largely complete by now, just waiting for approval from @prrace for the awt parts of the codebase. The vast majority of this is just hand-inlining macros in awt to avoid jumping across locals with gotos, it's very messy on the surface, sorry about that :(

And of course, as soon as I say that I get a mass JTReg failure that makes nearly every run on GHA turn red, thanks for the heart attack JTReg

TheShermanTanker commented:
@prrace ?

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:
```diff
@@ -6356,22 +6356,51 @@ jboolean AwtComponent::_NativeHandlesWheelScrolling(void *param)
 
-        AwtComponent *awtComponent = NULL;
-        AwtComponent *awtParent = NULL;
+        AwtComponent *awtComponent = nullptr;
```
Looking at it (not tested) here through 6403 could be simplified as

      if (self == NULL || parent == NULL) {
            env->ExceptionClear();
            JNU_ThrowNullPointerException(env, "peer");
            env->DeleteGlobalRef(self);
            env->DeleteGlobalRef(parent);
            delete data;
            return;
        }

        AwtComponent *awtComponent = (AwtComponent *)JNI_GET_PDATA(self);
        if (awtComponent == NULL) {
            THROW_NULL_PDATA_IF_NOT_DESTROYED(self);
            env->DeleteGlobalRef(self);
            env->DeleteGlobalRef(parent);
            delete data;
            return;
        }

        AwtComponent *awtParent = (AwtComponent *)JNI_GET_PDATA(parent);
        if (awtParent == NULL) {
            THROW_NULL_PDATA_IF_NOT_DESTROYED(parent);
            env->DeleteGlobalRef(self);
            env->DeleteGlobalRef(parent);
            delete data;
            return;
        }


I think I prefer that over
        BOOL error = FALSE;
        AwtComponent *awtComponent = nullptr;
        AwtComponent *awtParent = nullptr;

        if (self == NULL || parent == NULL) {
            env->ExceptionClear();
            JNU_ThrowNullPointerException(env, "peer");
            error = TRUE;
        }
        
        if (!error) {
            awtComponent = (AwtComponent *)JNI_GET_PDATA(self);
             if (awtComponent == NULL) {
                 THROW_NULL_PDATA_IF_NOT_DESTROYED(self);
                 error = TRUE;
             }
        }

        if (!error) {
            awtParent = (AwtComponent *)JNI_GET_PDATA(parent);
            if (awtParent == NULL) {
                THROW_NULL_PDATA_IF_NOT_DESTROYED(parent);
                error = TRUE;
        }

        if (error) {
            env->DeleteGlobalRef(self);
            env->DeleteGlobalRef(parent);
            delete data;
            return;
        }

TheShermanTanker replied:
Sorry, I don't see a BOOL error anywhere?

TheShermanTanker replied:
I see the advantage of collapsing self and parent into the same check, but it doesn't seem like getting rid of pData is of much benefit, the checks for null seem to remain the same either way

magicus replied:
> Sorry, I don't see a BOOL error anywhere?

I think Phil misplaced the code block marker -- the BOOL error line and below is supposed to be inside the second code block.

What Phil is doing is trying to suggest two different approaches to make this code nicer, the first with a bit more duplication, and the second with an `error` boolean flag. And he says that he prefers the former, so the second were propably just mentioned to show an alternative for discussion.

I think what you should do here is to combine the `self` and `parent` null checks as Phil suggests. And then replace all pData references to the final variable, here as well as everywhere pData is involved.

TheShermanTanker replied:
Ah, I see. Alright, will do so

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1336,22 +1336,33 @@ jobject AwtFrame::_GetBoundsPrivate(void *param)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtFrame::_SetState(void *param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
more un-needed reformatting

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1336,22 +1336,33 @@ jobject AwtFrame::_GetBoundsPrivate(void *param)
     HWND hwnd = f->GetHWnd();
-    if (::IsWindow(hwnd))
-    {
+    if (::IsWindow(hwnd)) {
```
more formatting to revert

magicus replied:
```suggestion
    if (::IsWindow(hwnd))
    {
```

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1567,26 +1578,63 @@ void AwtFrame::_SetIMMOption(void *param)
 void AwtFrame::_NotifyModalBlocked(void *param)
 {
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
again ..

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1567,26 +1578,63 @@ void AwtFrame::_SetIMMOption(void *param)
+        delete nmbs;
+        return;
+    } else {
+        pData = JNI_GET_PDATA(peer);
```
like in my recoded case above, we no longer need the "pData" var which was there only because that name
is magically known to the macro.

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp:
```diff
@@ -1567,26 +1578,63 @@ void AwtFrame::_SetIMMOption(void *param)
+        delete nmbs;
+        return;
+    } else {
+        pData = JNI_GET_PDATA(blockerPeer);
```
can directly set dialog

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -1014,21 +1014,32 @@ void AwtWindow::FocusedWindowChanged(HWND from, HWND to)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtWindow::_RepositionSecurityWarning(void* param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
formatting ..

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -1014,21 +1014,32 @@ void AwtWindow::FocusedWindowChanged(HWND from, HWND to)
+        return;
+    } else {
+        pData = JNI_GET_PDATA(self);
+        if (pData == NULL) {
```
directly set window

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3112,51 +3123,61 @@ void AwtWindow::_SetFocusableWindow(void *param)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtWindow::_ModalDisable(void *param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
formatting

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3112,51 +3123,61 @@ void AwtWindow::_SetFocusableWindow(void *param)
     PDATA pData = JNI_GET_PDATA(self);
-    if (pData == NULL) {
-        env->DeleteGlobalRef(self);
+    if (self == NULL) {
```
Surely line 3136 above should have been deleted ? It is replaced by line 3143 below.
And then you can also directly set window, pData isn't needed, as discussed in similar cases above

magicus replied:
Yes. 

@TheShermanTanker This is in fact a serious bug. You try to do `pData = JNI_GET_PDATA(self)` before the null check of self, and this will crash if self is null.

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3112,51 +3123,61 @@ void AwtWindow::_SetFocusableWindow(void *param)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtWindow::_ModalEnable(void *param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
fomatting

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3112,51 +3123,61 @@ void AwtWindow::_SetFocusableWindow(void *param)
     PDATA pData = JNI_GET_PDATA(self);
-    if (pData == NULL) {
-        env->DeleteGlobalRef(self);
+    if (self == NULL) {
```
like the method above line 3170 seems wrong.

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,64 +3186,102 @@ void AwtWindow::_ModalEnable(void *param)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtWindow::_SetOpaque(void* param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
formatting

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,64 +3186,102 @@ void AwtWindow::_ModalEnable(void *param)
+        return;
+    } else {
+        pData = JNI_GET_PDATA(self);
+        if (pData == NULL) {
```
set directly

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,64 +3186,102 @@ void AwtWindow::_ModalEnable(void *param)
-{
-    JNIEnv *env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
+void AwtWindow::_UpdateWindow(void* param) {
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
```
formatting

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,64 +3186,102 @@ void AwtWindow::_ModalEnable(void *param)
+        delete uws;
+        return;
+    } else {
+        pData = JNI_GET_PDATA(self);
```
set directly

prrace commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3240,12 +3299,24 @@ void AwtWindow::_SetFullScreenExclusiveModeState(void *param)
+        delete data;
+        return;
+    } else {
+        pData = JNI_GET_PDATA(self);
```
set directly

magicus replied:
@TheShermanTanker What this means is that you need to replace the line:
```
    PDATA pData;
```
above with:
```
 AwtWindow *window;
```
 
and remove 
```
    AwtWindow *window = (AwtWindow *)pData;
```
below, and then change all references to `pData` in this block to `window`.

The same changes needs to be done in all places where pData is references. This was a temporary variable that is no longer needed.

TheShermanTanker commented:
Fixed the formatting (at least in the marked cases), but am unsure what you mean by set directly?

prrace commented:
> Fixed the formatting (at least in the marked cases), but am unsure what you mean by set directly?

See my comment 
"like in my recoded case above, we no longer need the "pData" var which was there only because that name
is magically known to the macro."

so skip (and get rid of) the pData var and just set the target var directly

magicus commented:
bot-keep-alive

@TheShermanTanker Did you understand the remaining changes that Phil has requested? Do you think you'll be able to fix this?

TheShermanTanker commented:
> bot-keep-alive
> 
> @TheShermanTanker Did you understand the remaining changes that Phil has requested? Do you think you'll be able to fix this?

Hi Magnus, yes I do plan on fixing this. I've just been a bit busy and tired is all, and also to be honest I don't quite understand some of Phil's reviews, which is why I'm asking for clarification on some of them first

TheShermanTanker commented:
@prrace Sorry for the page, but as I mentioned above I don't see much of an advantage to removing pData, since it appears we still have to do the null check anyway even without it. I'm also not sure what the incorrect code you mentioned above is

TheShermanTanker commented:
@magicus  Phil doesn't seem to be responding to my queries, I'm not too sure what to do at this point

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_DnDDS.cpp:
```diff
@@ -39,10 +39,14 @@ void * operator new(size_t size) {return operator new(size, "stl", 1);}
-#include <awt.h>
 #include <shlobj.h>
 
+// Don't want to pull in the redefined allocation functions
```
Maybe this needs to be clarified a bit:

```suggestion
// These files must be included before awt.h, since the latter redefines malloc 
// to Do_Not_Use_Malloc, etc, and that will break these files.
```

Also, please mark conversations as resolved in the GitHub UI once you think they are addressed. That makes it much easier for reviewers to understand what parts of the reviews that have not been addressed.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_DnDDT.cpp:
```diff
@@ -27,11 +27,12 @@
 #include <shellapi.h>
 #include <memory.h>
 
+// Don't want to pull in the redefined allocation functions
+#include "awt_ole.h"
```
```suggestion
// awt_ole.h must be included before awt.h, since the latter redefines malloc 
// to Do_Not_Use_Malloc, etc, and that will break awt_ole.h.
#include "awt_ole.h"

```

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -540,12 +541,9 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
     */
-    else if (awtParent != NULL)
-    {
+    else if (awtParent != NULL) {
```
```suggestion
    else if (awtParent != NULL)
    {
```

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -540,12 +541,9 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
-    }
-    else
-    {
+    } else {
```
```suggestion
    }
    else
    {
```

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -708,7 +776,7 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
     }
 
-done:
+
```
```suggestion
```

magicus replied:
To be clear: I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -1050,16 +1269,51 @@ Java_sun_awt_windows_WPrinterJob_validatePaper(JNIEnv *env, jobject self,
     env->CallVoidMethod(newPaper, setImageableID, ix, iy, iw, ih);
 
-done:
+
```
```suggestion
```

magicus replied:
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3116,28 +3128,34 @@ void AwtWindow::_ModalDisable(void *param)
     }
 
-    window = (AwtWindow *)pData;
+    window = (AwtWindow *) pData;
```
```suggestion
    window = (AwtWindow *)pData;
```

TheShermanTanker commented:
Thanks for the clarifications! Makes my job a lot easier. The only thing I'm uncertain about is the pData local, which I don't see much benefit in removing since the null check associated with it still has to remain for code semantics to be correct

TheShermanTanker commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -1005,10 +1224,10 @@ Java_sun_awt_windows_WPrinterJob_validatePaper(JNIEnv *env, jobject self,
-    jdouble imgY = (jdouble)((yPixelOrg * 72)/(jdouble)yPixelRes);
-    jdouble imgWid = (jdouble)((imgPixelWid * 72)/(jdouble)xPixelRes);
-    jdouble imgHgt = (jdouble)((imgPixelHgt * 72)/(jdouble)yPixelRes);
+    jdouble imgX = (jdouble) ((xPixelOrg * 72)/(jdouble) xPixelRes);
+    jdouble imgY = (jdouble) ((yPixelOrg * 72)/(jdouble) yPixelRes);
+    jdouble imgWid = (jdouble) ((imgPixelWid * 72)/(jdouble) xPixelRes);
+    jdouble imgHgt = (jdouble) ((imgPixelHgt * 72)/(jdouble) yPixelRes);
```
I don't think I can commit this as there are 3 backticks at the end there :P

TheShermanTanker replied:
I know, I did want to commit your change directly though, just can't because of the backticks :P

magicus commented:
> The only thing I'm uncertain about is the pData local, which I don't see much benefit in removing since the null check associated with it still has to remain for code semantics to be correct

The point is that you can do the null check on the correct variable directly, instead of going a detour with pData. So instead of:

```
RealType realVal;
void* pData = getVal()
if (pData == null) {
  // bail out
}
realVal = (RealType) pData;
```

you can just do:

```
RealType realVal = getVal();
if (realVal == null) {
  // bail out
}
```

as the code normally would have been written, had there not been an old macro that used the "magic" temporary pData variable.

This is a recurring pattern in this patch and needs to be fixed everywhere.

TheShermanTanker commented:
> > The only thing I'm uncertain about is the pData local, which I don't see much benefit in removing since the null check associated with it still has to remain for code semantics to be correct
> 
> The point is that you can do the null check on the correct variable directly, instead of going a detour with pData. So instead of:
> 
> ```
> RealType realVal;
> void* pData = getVal()
> if (pData == null) {
>   // bail out
> }
> realVal = (RealType) pData;
> ```
> 
> you can just do:
> 
> ```
> RealType realVal = getVal();
> if (realVal == null) {
>   // bail out
> }
> ```
> 
> as the code normally would have been written, had there not been an old macro that used the "magic" temporary pData variable.
> 
> This is a recurring pattern in this patch and needs to be fixed everywhere.

I have a concern since the null check bailout involves THROW\_NULL\_PDATA\_IF\_NOT\_DESTROYED, which is no longer accurate if we remove the pData local. Maybe Phil can suggest an alternative to that, because I don't know what that might be. Although, if given the choice I'd use RAII through unique\_ptr like Thomas suggested, it just seems so much cleaner to me. Speaking of, hopefully this causes enough noise to get Phil's attention again. In the meantime, I've reverted as much of the formatting changes I could possibly find

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Canvas.cpp:
```diff
@@ -206,19 +206,31 @@ void AwtCanvas::_SetEraseBackground(void *param)
+    }
 
-    AwtCanvas *c = (AwtCanvas*)pData;
+    AwtCanvas *c = (AwtCanvas*) pData;
```
```suggestion
    AwtCanvas *c = (AwtCanvas*)pData;
```

magicus commented:
> I have a concern since the null check bailout involves THROW_NULL_PDATA_IF_NOT_DESTROYED, which is no longer accurate if we remove the pData local.

The name of the macro is not great, but it does not involve pData (the bad NPE error message notwithstanding):

```
#define THROW_NULL_PDATA_IF_NOT_DESTROYED(peer) {                         \
    jboolean destroyed = JNI_GET_DESTROYED(peer);                         \
    if (destroyed != JNI_TRUE) {                                          \
        env->ExceptionClear();                                            \
        JNU_ThrowNullPointerException(env, "null pData");                 \
    }                                                                     \
}
```

TheShermanTanker commented:
> > I have a concern since the null check bailout involves THROW_NULL_PDATA_IF_NOT_DESTROYED, which is no longer accurate if we remove the pData local.
> 
> The name of the macro is not great, but it does not involve pData (the bad NPE error message notwithstanding):
> 
> ```
> #define THROW_NULL_PDATA_IF_NOT_DESTROYED(peer) {                         \
>     jboolean destroyed = JNI_GET_DESTROYED(peer);                         \
>     if (destroyed != JNI_TRUE) {                                          \
>         env->ExceptionClear();                                            \
>         JNU_ThrowNullPointerException(env, "null pData");                 \
>     }                                                                     \
> }
> ```
> 
> So you can go ahead and replace the pData references with the variable that will eventually be used.

Alright, will do. Maybe as a further improvement, I can inline THROW\_NULL\_PDATA\_IF\_NOT\_DESTROYED at its callsites and replace the bad NullPointerException error message with the proper null pointer name. Since Phil isn't here, what do you think?

Regardless, I really hope I can get this in by Thursday. University for me officially ramps up into _very_ high gear about that time, and I doubt I can juggle both JDK work and it all at once by then

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -572,19 +573,29 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
         int sz = GetLocaleInfo(LOCALE_USER_DEFAULT, LOCALE_IMEASURE, NULL, 0);
         if (sz > 0) {
-          LPTSTR str = (LPTSTR)SAFE_SIZE_ARRAY_ALLOC(safe_Malloc, sizeof(TCHAR), sz);
+          LPTSTR str = (LPTSTR) SAFE_SIZE_ARRAY_ALLOC(safe_Malloc, sizeof(TCHAR), sz);
```
```suggestion
          LPTSTR str = (LPTSTR)SAFE_SIZE_ARRAY_ALLOC(safe_Malloc, sizeof(TCHAR), sz);
```

magicus commented:
> Maybe as a further improvement, I can inline THROW_NULL_PDATA_IF_NOT_DESTROYED at its callsites and replace the bad NullPointerException error message with the proper null pointer name. Since Phil isn't here, what do you think?

That might be a good future code cleanup, but I suggest you refrain from doing that in this change. 

> Regardless, I really hope I can get this in by Thursday. 

I see that you are struggling a bit to get this done. I can probably help you fix the remaining issues, but I can't push to your PR. So either I'd need to create a fork of your `patch-10` branch in my personal fork and you'll have to pull from there into your branch, or you need to give me write permissions to your personal fork. I believe you can assign permissions on a per-branch basis.

I'd really like to see this PR to come to conclusion.

TheShermanTanker commented:
Thanks, will assign the permissions. Though, I'll try to finish this myself so you don't have to\- It seems like bad etiquette to make you fix something that I started. But all of this will still be meaningless if Phil doesn't return to this Pull Request any time soon :(

magicus commented:
Let's fix all the known problems first, and he'll return. (Otherwise I'll ping him in a different channel.)

TheShermanTanker commented:
I've switched the marked cases to use direct setting and bypass pData, should I do the unmarked ones too? @prrace

TheShermanTanker commented:
Bumping

prrace commented:
I've been monitoring it but saw no reason to pay particular attention until everything I raised is addressed.
Including the cases identical to the ones I explicitly identified
(which I suppose is what is meant by your question "should I do the unmarked ones too ?").
At some point you do tire of typing "ditto" :-)

It has been long enough that I will need a little time to remember and re-review properly once the requested changes are in place.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:
```diff
@@ -6539,19 +6558,32 @@ static void _GetInsets(void* param)
     component->GetInsets(gis->insets);
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:
```diff
@@ -209,7 +211,6 @@ static const double POINTS_TO_HIMETRIC = (2540.0 / 72.0);
 static const double POINTS_TO_LOMETRIC = (254.0 / 72.0);
 
-jfieldID AwtPrintDialog::pageID;
 
```
I recommend deleting one more line, since otherwise you will now have three consecutive blank lines.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -1019,16 +1019,28 @@ void AwtWindow::_RepositionSecurityWarning(void* param)
 
     window->RepositionSecurityWarning(env);
 
-  ret:
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3137,7 +3156,7 @@ void AwtWindow::_ModalDisable(void *param)
     }
 
-ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,25 +3190,38 @@ void AwtWindow::_ModalEnable(void *param)
     }
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3165,25 +3190,38 @@ void AwtWindow::_ModalEnable(void *param)
     window->SetTranslucency(iOpacity, window->isOpaque());
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3192,17 +3230,29 @@ void AwtWindow::_SetOpaque(void* param)
     window->SetTranslucency(window->getOpacity(), isOpaque);
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3211,18 +3261,36 @@ void AwtWindow::_UpdateWindow(void* param)
                          uws->hBitmap);
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented on src/java.desktop/windows/native/libawt/windows/awt_Window.cpp:
```diff
@@ -3239,13 +3307,25 @@ void AwtWindow::_SetFullScreenExclusiveModeState(void *param)
     window->setFullScreenExclusiveModeState(state != 0);
 
-  ret:
+
```
I recommend deleting this line, since it does not make sense to have two consecutive blank lines here.

magicus commented:
@TheShermanTanker I tried to help you get this done. I added fixes to a copy of your branch on my personal fork, but then it turned out I could not push them to your branch. :-(

It ended up with me creating a new PR, https://github.com/openjdk/jdk/pull/18584. As a bonus, I think it might be easier to review with a fresh start. This PR has grown quite heavy with lots of comments and commits.

I hope you don't feel like I'm stealing this away from you. You have done a great job, and shown a lot of patience of carrying this all the way here. But I also got the impression that you would appreciate my assistance in getting the last pieces in place so we can integrate this.

TheShermanTanker commented:
> @TheShermanTanker I tried to help you get this done. I added fixes to a copy of your branch on my personal fork, but then it turned out I could not push them to your branch. :-(
> 
> It ended up with me creating a new PR, #18584. As a bonus, I think it might be easier to review with a fresh start. This PR has grown quite heavy with lots of comments and commits.
> 
> I hope you don't feel like I'm stealing this away from you. You have done a great job, and shown a lot of patience of carrying this all the way here. But I also got the impression that you would appreciate my assistance in getting the last pieces in place so we can integrate this.

Not at all, I don't feel like you're stealing this from me. In fact, I should be the one apologising for giving you extra work! Thanks for taking this up, I once again apologise for making you do this instead, I've been very busy since Thursday (working on OpenJDK while in lectures at times), and during my breaks I've been too drained to continue, so i really appreciate your help :)

I will keep this open until the other Pull Request has been integrated, in case this might still be needed

magicus commented:
> I will keep this open until the other Pull Request has been integrated, in case this might still be needed

I don't think it is necessary. You can always re-open a PR if it should be needed, and you can look at source code and comments (and even make new comments) on a closed PR. So I think it will just make it easier for everybody if there is only a single open PR for this issue.

magicus commented:
@TheShermanTanker You can close this PR. The bug was fixed with https://github.com/openjdk/jdk/pull/18584 instead.
